### PR TITLE
ROCm: Fix device attributes; Record HIP_VERSION; Add more device properties; etc 

### PR DESCRIPTION
**Pull request by leofang**
Main changes:
- Record `HIP_VERSION`, and report it in `cupy.show_config()`
- Compute Capability is meaningless in ROCm/HIP, so we instead report the arch in `cupy.show_config()`
- Use both `CUDA_VERSION` and `HIP_VERSION` as a mutex to indicate the build platform; if both =0 then it's for Read the Docs (RTD). This eliminates the need of passing an extra Cython macro `use_hip`.
- Fix device attributes: It turns out that the enum is defined differently on CUDA and HIP...
- Update device properties for a few fields added since ROCm 3.10
- Put fields in `hipDeviceProp.arch` in a dict to align with the C struct layout
- Don't pass a silly option like `-arch=compute_gfx906` to hipRTC/hipcc. Instead, pass `-arch=gfc906`, though it seems to me `-arch` was neglected...(maybe check with AMD people?)

Before:
```
OS                           : Linux-5.8.0-36-generic-x86_64-with-debian-bullseye-sid
CuPy Version                 : 9.0.0b1
NumPy Version                : 1.19.4
SciPy Version                : 1.6.0
Cython Build Version         : 0.29.21
Cython Runtime Version       : 0.29.21
CUDA Root                    : /opt/rocm
CUDA Build Version           : 0
CUDA Driver Version          : 321200
CUDA Runtime Version         : 3212
cuBLAS Version               : 23200
cuFFT Version                : 10008966
cuRAND Version               : 201006
cuSOLVER Version             : (3, 10, 0)
cuSPARSE Version             : 0
NVRTC Version                : (9, 0)
Thrust Version               : 100908
CUB Build Version            : 201005
Jitify Build Version         : None
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : 2708
NCCL Runtime Version         : 2708
cuTENSOR Version             : None
Device 0 Name                : Vega 20
Device 0 Compute Capability  : 90
```
After:
```
OS                      : Linux-5.8.0-36-generic-x86_64-with-debian-bullseye-sid
CuPy Version            : 9.0.0b1
NumPy Version           : 1.19.4
SciPy Version           : 1.6.0
Cython Build Version    : 0.29.21
Cython Runtime Version  : 0.29.21
CUDA Root               : /opt/rocm
CUDA Build Version      : 400
CUDA Driver Version     : 321200
CUDA Runtime Version    : 3212
cuBLAS Version          : 23200
cuFFT Version           : 10008966
cuRAND Version          : 201006
cuSOLVER Version        : (3, 10, 0)
cuSPARSE Version        : 0
NVRTC Version           : (9, 0)
Thrust Version          : 100908
CUB Build Version       : 201005
Jitify Build Version    : None
cuDNN Build Version     : None
cuDNN Version           : None
NCCL Build Version      : 2708
NCCL Runtime Version    : 2708
cuTENSOR Version        : None
Device 0 Name           : Vega 20
Device 0 Arch           : gfx906
```

### CODE REVIEW

**Comment by leofang** [comment]
@emcastillo The PR description is added. PTAL.

File: cupy/cuda/device.pyx [hunk 1]
```diff
@@ -73,12 +73,6 @@ def _get_attributes(device_id):
     d = {}
     for k, v in runtime_module.__dict__.items():
         if k.startswith('cudaDevAttr'):
-            if runtime._is_hip_environment:
-                # On ROCm 3.5.0 + gfx906, accessing these attributes leads to
-                # core dump (stack smashing detected)
-                if k in ('cudaDevAttrPciDeviceId',
-                         'cudaDevAttrTccDriver'):
-                    continue
```
Comment (leofang) [inline cupy/cuda/device.pyx:81]: So it turns out this bug we hit is due to the enum mismatch between CUDA and HIP. After fixing it, the actual problematic attributes are `hipDeviceAttributeHdpMemFlushCntl` and `hipDeviceAttributeHdpRegFlushCntl`, see my comments in `cupy_backends/cuda/api/runtime.pxd`. The fix is shifted there (by not exposing them to Python).

File: cupy/cusolver.py [hunk 2]
```diff
@@ -49,9 +50,7 @@ def check_availability(name):
         version = _runtime.runtimeGetVersion()
     else:
         available_version = _available_hip_version
-        # TODO(leofang): use HIP_VERSION instead?
-        version = _cusolver._getVersion()
-        version = version[0] * 100 + version[1]
+        version = _driver.get_build_version()  # = HIP_VERSION
```
Comment (leofang) [inline cupy/cusolver.py:53]: rocSOLVER's owns version is not always the same as `HIP_VERSION`.

File: cupy_backends/cuda/api/runtime.pxd [hunk 3]
```diff
@@ -613,9 +572,244 @@ cpdef enum:
     cudaReadModeNormalizedFloat = 1
 
 
+# This was a legacy mistake: the prefix "cuda" should have been removed
+# so that we can directly assign their C counterparts here. Now because
+# of backward compatibility and no flexible Cython macro (IF/ELSE), we
+# have to duplicate the enum. (CUDA and HIP use different values!)
```
Comment (leofang) [inline cupy_backends/cuda/api/runtime.pxd:578]: I am thinking whether we should rectify this mistake or not. For example, the attributes should be exposed under the same naming convention as other stuff in this module: `devAttr...`, with the prefix `cuda` removed and the following character being lowercase. Perhaps for backward compatibility we can make both the attributes (with old/new names) coexist, and in one or two releases we remove the old names?

**Comment by leofang** [comment]
Power-cycle to reboot RTD CI.

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy_backends/cuda/api/runtime.pxd (hunk 3 above)
Comment (emcastillo) [reply cupy_backends/cuda/api/runtime.pxd:578]: I think deprecating and removing is better

Comment (leofang) [reply cupy_backends/cuda/api/runtime.pxd:578]: @emcastillo Do you know how to emit warning if module-level attributes like these are accessed?

Comment (leofang) [reply cupy_backends/cuda/api/runtime.pxd:578]: The best solution I found is to utilize PEP 562, write a module-level `__getattr__()`, and emit a warning there. But it's Python 3.7+ only. Let's not worry about it now and do this in the next PR, as I have a second thought on the `cuda` prefix for runtime enums...😅

**Comment by leofang** [comment]
@emcastillo This is ready.

**Comment by leofang** [comment]
Jenkins, test this please

**Review by emcastillo — APPROVED** [review]

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Thanks, @emcastillo! Looks like mergify is blocked by the Windows CI.

**Comment by emcastillo** [comment]
Let's get it in then!

File: cupy_backends/cuda/api/runtime.pxd [hunk 4]
```diff
@@ -469,117 +530,6 @@ cpdef enum:
...
-    cudaDevAttrReserved93 = 93
-    cudaDevAttrReserved94 = 94
-    cudaDevAttrCooperativeLaunch = 95
-    cudaDevAttrCooperativeMultiDeviceLaunch = 96
-    cudaDevAttrMaxSharedMemoryPerBlockOptin = 97
-    cudaDevAttrCanFlushRemoteWrites = 98
-    cudaDevAttrHostRegisterSupported = 99
-    cudaDevAttrPageableMemoryAccessUsesHostPageTables = 100
-    cudaDevAttrDirectManagedMemAccessFromHost = 101
-    # added since CUDA 11.0
```
Comment (kmaehashi) [inline cupy_backends/cuda/api/runtime.pxd:573]: @leofang These enum lost in this PR, causing test failures in CUDA 11.2

File: cupy_backends/cuda/api/runtime.pxd [hunk 5]
```diff
@@ -469,117 +530,6 @@ cpdef enum:
...
-    cudaDevAttrMaxBlocksPerMultiprocessor = 106
-    cudaDevAttrReservedSharedMemoryPerBlock = 111
-    # added since CUDA 11.1
-    cudaDevAttrSparseCudaArraySupported = 112
-    cudaDevAttrHostRegisterReadOnlySupported = 113
-    # added since CUDA 11.2
-    cudaDevAttrMaxTimelineSemaphoreInteropSupported = 114
-    cudaDevAttrMemoryPoolsSupported = 115
```
Comment (kmaehashi) [inline cupy_backends/cuda/api/runtime.pxd:581]: @leofang These enum lost in this PR, causing test failures in CUDA 11.2

Comment (kmaehashi) [reply cupy_backends/cuda/api/runtime.pxd:581]: > E   AttributeError: module 'cupy_backends.cuda.api.runtime' has no attribute 'cudaDevAttrMemoryPoolsSupported'

Comment (leofang) [reply cupy_backends/cuda/api/runtime.pxd:581]: shoot, let me send a pr @kmaehashi

**Pull request merged**